### ISSUE TITLE
tf.contrib.data.Dataset  does not handle well with last  elements with is fewer than batch size

### ISSUE THREAD

**Issue by zh794390558** [tensorflow/tensorflow#13745]
tf.contrib.data.Dataset  does not handle well with last  elements with is fewer than batch size.

Maybe batch_size = 10, but last batch has 9 elements.

**Comment by mrry** [comment]
When this is that case `dataset.batch(10)` will give you zero or more batches of 10 elements, followed by a single batch of 9 elements. If you don't want the smaller batch, you can use `dataset.apply(tf.contrib.data.batch_and_drop_remainder(10))`. 

What would you recommend instead?

**Comment by zh794390558** [comment]
I think which may have a param for dataset.batch()  which like allow_smaller_final_batch of tf.train.batch().

**Comment by mrry** [comment]
If you're using `Dataset.padded_batch()`, you can compose it with a `Dataset.filter()` to exclude smaller batches. For example, if your dataset has three components in each element `(x, y, z)`:

```python
dataset = dataset.padded_batch(batch_size, ...).filter(lambda x, y, z: tf.equal(tf.shape(x), batch_size))
```

**Comment by zh794390558** [comment]
@mrry Thanks, it's work.

**Comment by glhfgg1024** [comment]
Hi, if you are convenient, please help check the following scripts. I use the `cifar10` datasets. The `train` subset has 40000 samples, the `val` subset has 10000 samples and the `test` subset has 10000 samples. 
The following script is to generate the tfrecords for cifar10 dataset with a `shard_size` parameter, which will generate multiple `tfrecords` files for one subset, e.g., `trn-001.tfrecords`, `trn-002.tfrecords`, `trn-003.tfrecords`, etc. 
        
    from __future__ import absolute_import
    from __future__ import division
    from __future__ import print_function
    
    import argparse
    import os
    
    import tarfile
    from six.moves import cPickle as pickle
    from six.moves import xrange  # pylint: disable=redefined-builtin
    import tensorflow as tf
    import numpy as np
    
    CIFAR_FILENAME = 'cifar-10-python.tar.gz'
    CIFAR_DOWNLOAD_URL = 'https://www.cs.toronto.edu/~kriz/' + CIFAR_FILENAME
    CIFAR_LOCAL_FOLDER = 'cifar-10-batches-py'
    
    
    def download_and_extract(data_dir):
      # download CIFAR-10 if not already downloaded.
      tf.contrib.learn.datasets.base.maybe_download(CIFAR_FILENAME, data_dir,
                                                    CIFAR_DOWNLOAD_URL)
      tarfile.open(os.path.join(data_dir, CIFAR_FILENAME),
                   'r:gz').extractall(data_dir)
    
    
    def _int64_feature(value):
      return tf.train.Feature(int64_list=tf.train.Int64List(value=[value]))
    
    
    def _bytes_feature(value):
      return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))
    
    
    def _get_file_names():
      """Returns the file names expected to exist in the input_dir."""
      file_names = {}
      file_names['trn'] = ['data_batch_%d' % i for i in xrange(1, 5)]
      file_names['val'] = ['data_batch_5']
      file_names['tst'] = ['test_batch']
      return file_names
    
    
    def read_pickle_from_file(filename):
      with tf.gfile.Open(filename, 'rb') as f:
        data_dict = pickle.load(f)
      return data_dict
    
    
    def convert_to_tfrecord(input_files, data_dir, mode):
        """Converts a file to TFRecords."""
        data_dict = read_pickle_from_file(input_files[0])
        data = data_dict['data']
        labels = data_dict['labels']
        print(data.shape)
        for ii in range(1,len(input_files)):
            data_dict = read_pickle_from_file(input_files[ii])
            data = np.concatenate([data, data_dict['data']], axis=0)
            labels = labels + data_dict['labels']
    
        print(data.shape, len(labels))
        shard_size = 8192
        print(len(labels))
        num_batches = int(np.ceil(len(labels)/shard_size))
        for bi in range(num_batches):
            output_file = '{}/{}_{:03d}.tfrecords'.format(data_dir, mode, bi)
            s0 = bi*shard_size
            s1 = (bi+1)*shard_size
            if s1>len(labels):
                s1 = len(labels)
            print(mode, bi, s0, s1)
            with tf.python_io.TFRecordWriter(output_file) as record_writer:
                for i in range(s0,s1,1):
                    example = tf.train.Example(features=tf.train.Features(
                        feature={
                            'image': _bytes_feature(data[i].tobytes()),
                            'label': _int64_feature(labels[i])
                        }))
                    record_writer.write(example.SerializeToString())
    
    
    def main(data_dir):
      print('Download from {} and extract.'.format(CIFAR_DOWNLOAD_URL))
      download_and_extract(data_dir)
      file_names = _get_file_names()
      input_dir = os.path.join(data_dir, CIFAR_LOCAL_FOLDER)
      for mode, files in file_names.items():
        input_files = [os.path.join(input_dir, f) for f in files]
        # Convert to tf.train.Example and write the to TFRecords.
        convert_to_tfrecord(input_files, data_dir, mode)
      print('Done!')
    
    
    if __name__ == '__main__':
      parser = argparse.ArgumentParser()
      parser.add_argument(
          '--data-dir',
          type=str,
          default='./data',
          help='Directory to download and extract CIFAR-10 to.')
    
      args = parser.parse_args()
    main(args.data_dir)

After generating the sharded `tfrecords` files for the three subsets. I use the following script to compute the mean values in `train + val` subsets, which in total have `40000+10000=50000` samples. The `Dataset` API is used to provide batch inputs. Here, for testing, I set the batch size as `128`, which means we have in total `50000/128=390`  full batches and `50000%128=80` sample in the last batch. Owing to the `dataset = dataset.apply(tf.contrib.data.batch_and_drop_remainder(batch_size))` function, it is expected to drop the last batch. But this script didn't work, the error information is provided in the end.

    from __future__ import unicode_literals
    from __future__ import print_function
    from __future__ import division
    from __future__ import absolute_import
    
    import tensorflow as tf
    import os
    import numpy as np
    import pandas as pd
    import glob
    import argparse
    import sklearn
    import sklearn.metrics
    
    
    HEIGHT = 32
    WIDTH = 32
    DEPTH = 3
    
    def dataset_input_from_tfrecords(filenames, batch_size=1, num_epochs=1, shuffle=True, image_mean=None):
        # filenames = ["/var/data/file1.tfrecord", "/var/data/file2.tfrecord"]
        dataset = tf.data.TFRecordDataset(filenames)
        
        # Use `tf.parse_single_example()` to extract data from a `tf.Example`
        # protocol buffer, and perform any additional per-record preprocessing.
        def parser(record):
            keys_to_features = {
                'image': tf.FixedLenFeature([], tf.string),
                'label': tf.FixedLenFeature([], tf.int64),
            }
            parsed = tf.parse_single_example(record, keys_to_features)
            
            # Perform additional preprocessing on the parsed data.
            image = tf.cast(tf.decode_raw(parsed['image'], tf.uint8), tf.float32)/128 - 1.0
            image = tf.transpose(tf.reshape(image, [DEPTH, HEIGHT, WIDTH]), [1,2,0])
            label = tf.cast(parsed['label'], tf.int64)
            
            if image_mean != None:
                image -= image_mean
                
            if shuffle:
                image = tf.image.resize_image_with_crop_or_pad(image, 40, 40)
                image = tf.random_crop(image, [HEIGHT, WIDTH, DEPTH])
                image = tf.image.random_flip_left_right(image)
            
            return image, label
        
        # Use `Dataset.map()` to build a pair of a feature dictionary and a label 
        # tensor for each example.
        dataset = dataset.map(parser)
        if shuffle:
            dataset = dataset.shuffle(buffer_size=256)
        dataset = dataset.batch(batch_size)
        **dataset = dataset.apply(tf.contrib.data.batch_and_drop_remainder(batch_size))**
        dataset = dataset.repeat(num_epochs)
        iterator = dataset.make_one_shot_iterator()
        
        features, labels = iterator.get_next()
        
        return features, labels
    
    trn_filenames = glob.glob('data/trn*.tfrecords')
    val_filenames = glob.glob('data/val*.tfrecords')
    tst_filenames = glob.glob('data/tst*.tfrecords')
    print(len(trn_filenames), len(val_filenames), len(tst_filenames))
    
    
    def compute_mean_std():
        
        batch_size=128
        **filenames = trn_filenames + val_filenames**
        features_op, labels_op = dataset_input_from_tfrecords(filenames, batch_size=batch_size, num_epochs=1, shuffle=False)
            
        config = tf.ConfigProto()
        config.gpu_options.allow_growth = True 
        sess = tf.Session(config=config)
            
        sess.run(tf.global_variables_initializer())
        sess.run(tf.local_variables_initializer())
        
        coord = tf.train.Coordinator()
        threads = tf.train.start_queue_runners(sess=sess, coord=coord)
        
        count = 0
        try:
            while not coord.should_stop():
                features, labels = sess.run([features_op, labels_op])
                if count%2000==0:
                    print(features.shape, labels.shape)
                count += batch_size
        except tf.errors.OutOfRangeError:
            print('done.')
        finally:
            coord.request_stop()
            print(count)
            
        coord.join(threads)
        sess.close()
    
And the `compute_mean_std()` will have the following errors:

    InvalidArgumentError (see above for traceback): Cannot batch tensors with different shapes in component 0. First element had shape [128,32,32,3] and element 6 had shape [80,32,32,3].
	 [[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[128,?,32,32,3], [128,?]], output_types=[DT_FLOAT, DT_INT64], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator)]]

**Comment by chenghuige** [comment]
What if I want to extend final batch to 10 elements from 9 by extending one element from existing 9 ?

**Comment by mrry** [comment]
@chenghuige You can handle that by adding a `Dataset.map()` that uses `tf.pad()` to pad each batch up to 10 elements.

**Comment by lixiang-ucas** [comment]
@mrry can you give an example of extending one element from existing 9? What if my input tensor has a shape of [batch_size,10,3]?
Thank you.

**Comment by NikkiZy** [comment]
I have met the same problem, that is what i tried:
dataset = dataset.padded_batch(....., drop_reminder = True) 
dataset = dataset.repeat()
and it works for me. Good luck

**Comment by JorgeCeja** [comment]
`tf.contrib.data.batch_and_drop_remainder()` is DEPRECATED

Use `tf.data.Dataset.batch(..., drop_remainder=True)` INSTEAD

**Comment by tsvetiko** [comment]
As @mrry said, I needed to do the following (considering the shape of the labels) in order to work:
```
dataset = dataset.batch(batch_size).filter(lambda features, labels: tf.equal(tf.shape(labels)[0], batch_size))
```

**Comment by LaCandela** [comment]
> I have met the same problem, that is what i tried:
> dataset = dataset.padded_batch(....., drop_reminder = True)
> dataset = dataset.repeat()
> and it works for me. Good luck

Thanks, this works

**Comment by wilderrodrigues** [comment]
It's year 2020 now and I faced the same issue. It was hard to understand because there were many loose parts. For instance, we have 4 TF records and I was loading them, preprocessing because we had issues with some labels, converting to tf.data.Dataset, concatenating them and saving the shards with the new `save` function from TF2.3.0rc2. So, got really lost in the code and forgot to look at the batch. The issue was also only happening when I was adding the augmentation, after loading the Datasets before training.

Took me sometime to remember that the augmentation was batching it and having some remaining data that was not inline with the expected shapes. So, the `padded_patch(ds, drop_remainder=True).repeat()` fixed it. 👍
